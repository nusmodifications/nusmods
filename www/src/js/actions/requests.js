// @flow

import type { AxiosXHRConfig } from 'axios';
import type { FSA } from 'types/redux';
import type { RequestKey } from 'types/reducers';

export const API_REQUEST = 'API_REQUEST';

export type RequestAction = FSA & {
  payload: AxiosXHRConfig<*>,
  meta: {
    [typeof API_REQUEST]: string,
  },
};

/* eslint-disable import/prefer-default-export, no-param-reassign */
type RequestActionCreator = {
  (key: RequestKey, options: AxiosXHRConfig<*>): RequestAction,
  (key: RequestKey, type: string, options: AxiosXHRConfig<*>): RequestAction,
};

/**
 * Create an action that makes a HTTP request. key is a string that uniquely
 * identifies each request. If type is not provided then it is assumed to be
 * the same as key. Options is passed directly into axios config.
 */
export const requestAction: RequestActionCreator = (key, type, options) => {
  let payload: AxiosXHRConfig<*>;

  if (typeof type !== 'string') {
    payload = type;
    type = key;
  } else if (options) {
    payload = options;
  } else {
    // Keeps Flow happy
    throw new Error();
  }

  return {
    type,
    payload,
    meta: {
      [API_REQUEST]: key,
    },
  };
};
