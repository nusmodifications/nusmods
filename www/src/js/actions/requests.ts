// @flow

import type { AxiosXHRConfig } from 'axios';
import type { RequestKey } from 'types/reducers';

export const API_REQUEST = 'API_REQUEST';

export type RequestAction = {|
  type: string,
  payload: AxiosXHRConfig<*>,
  meta: {
    [typeof API_REQUEST]: string,
  },
|};

/* eslint-disable import/prefer-default-export, no-param-reassign */
type RequestActionCreator = {
  (key: RequestKey, options: AxiosXHRConfig<*>): RequestAction,
  (key: RequestKey, type: string, options: AxiosXHRConfig<*>): RequestAction,
};

/**
 * Create an action that makes a HTTP request. key is a string that uniquely
 * identifies each request. If type is not provided then it is assumed to be
 * the same as key.
 *
 * For example, fetching individual module info uses the type
 * FETCH_MODULE, while the key is FETCH_MODULE_[Module Code]. This allows the
 * status of each individual request to be stored, while the actions are still
 * grouped by their type.
 *
 * Options is passed directly into axios config, so minimally url should be
 * provided.
 */
export const requestAction: RequestActionCreator = (key, type, options) => {
  let payload: AxiosXHRConfig<*>;

  if (typeof type !== 'string') {
    payload = type;
    type = key;
  } else if (options) {
    payload = options;
  } else {
    // Keeps Flow happy
    throw new Error();
  }

  return {
    type,
    payload,
    meta: {
      [API_REQUEST]: key,
    },
  };
};
