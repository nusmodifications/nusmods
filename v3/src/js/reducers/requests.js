import _ from 'lodash';

import * as helperActions from 'actions/helpers';
import * as requestResultCases from 'middlewares/requests-middleware';
import type { FetchRequest } from 'types/reducers';

const requestState = 'Request';

export default function requests(state = {}, action = null): FetchRequest {
  const { type, requestStatus } = action;

  if (type === helperActions.RESET_ALL_STATE) {
    return {};
  }

  // Reset request state
  if (type === helperActions.RESET_REQUEST_STATE) {
    const newState = {};
    _.each(action.payload, (domain) => {
      newState[_.camelCase(domain) + requestState] = {
        isPending: false,
        isSuccessful: false,
        isFailure: false,
      };
    });

    return _.merge({}, state, newState);
  }

  // requestStatus is a field specially designed and owned by api request actions
  let domain = '';
  switch (requestStatus) {
    case requestResultCases.REQUEST:
      domain = _.camelCase(type.replace(requestResultCases.REQUEST, ''));
      return _.merge({}, state, {
        [domain + requestState]: {
          isPending: true,
          isSuccessful: false,
          isFailure: false,
        },
      });
    case requestResultCases.SUCCESS:
      domain = _.camelCase(type.replace(requestResultCases.SUCCESS, ''));
      return _.merge({}, state, {
        [domain + requestState]: {
          isPending: false,
          isSuccessful: true,
          isFailure: false,
        },
      });
    case requestResultCases.FAILURE:
      domain = _.camelCase(type.replace(requestResultCases.FAILURE, ''));
      return _.merge({}, state, {
        [domain + requestState]: {
          isPending: false,
          isSuccessful: false,
          isFailure: true,
        },
      });
    default:
      return state;
  }
}
