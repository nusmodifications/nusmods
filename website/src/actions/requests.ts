import { AxiosRequestConfig } from 'axios';
import { RequestKey } from 'types/reducers';

export const API_REQUEST = 'API_REQUEST';

export type RequestAction = {
  type: string;
  payload: AxiosRequestConfig;
  meta: {
    [key: string]: string;
  };
};

/**
 * Create an action that makes a HTTP request. key is a string that uniquely
 * identifies each request. If type is not provided then it is assumed to be
 * the same as key.
 *
 * For example, fetching individual module info uses the type
 * FETCH_MODULE, while the key is FETCH_MODULE_[Module Code]. This allows the
 * status of each individual request to be stored, while the actions are still
 * grouped by their type.
 *
 * Options is passed directly into axios config, so minimally url should be
 * provided.
 */
type RequestActionCreator = {
  (key: RequestKey, options: AxiosRequestConfig): RequestAction;
  (key: RequestKey, type: string, options: AxiosRequestConfig): RequestAction;
};

export const requestAction: RequestActionCreator = (
  key: RequestKey,
  type: string | AxiosRequestConfig,
  options?: AxiosRequestConfig,
): RequestAction => {
  let payload: AxiosRequestConfig;

  if (typeof type !== 'string') {
    payload = type;
    // eslint-disable-next-line no-param-reassign
    type = key;
  } else if (options) {
    payload = options;
  } else {
    // Keeps Flow happy
    throw new Error();
  }

  return {
    type,
    payload,
    meta: {
      [API_REQUEST]: key,
    },
  };
};
