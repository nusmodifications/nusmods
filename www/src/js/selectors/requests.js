// @flow
import type { State } from 'reducers';
import { SUCCESS, FAILURE, REQUEST } from 'types/reducers';

export function isOngoing(state: State, key: string) {
  return state.requests[key] && state.requests[key].status === REQUEST;
}

export function isSuccess(state: State, key: string) {
  return state.requests[key] && state.requests[key].status === SUCCESS;
}

export function isFailure(state: State, key: string) {
  return state.requests[key] && state.requests[key].status === FAILURE;
}
