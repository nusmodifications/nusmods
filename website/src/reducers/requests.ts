import { FSA } from 'types/redux';
import { Requests, FAILURE, REQUEST, SUCCESS } from 'types/reducers';
import { API_REQUEST } from 'actions/requests';

export default function requests(state: Requests = {}, action: FSA): Requests {
  const { meta } = action;

  // requestStatus is a field specially designed and owned by api request actions
  if (!meta || !meta.requestStatus || !meta[API_REQUEST]) {
    return state;
  }

  const key = meta[API_REQUEST];

  switch (meta.requestStatus) {
    case REQUEST:
      return {
        ...state,
        [key]: {
          status: REQUEST,
        },
      };

    case SUCCESS:
      return {
        ...state,
        [key]: {
          status: SUCCESS,
        },
      };

    case FAILURE:
      return {
        ...state,
        [key]: {
          status: FAILURE,
          error: action.payload,
        },
      };

    default:
      return state;
  }
}
