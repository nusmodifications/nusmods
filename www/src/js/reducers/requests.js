// @flow
import { camelCase } from 'lodash';

import type { FSA } from 'types/redux';
import type { Requests, RequestType, FetchRequest } from 'types/reducers';

import { RESET_ALL_STATE, RESET_REQUEST_STATE } from 'actions/helpers';
import { REQUEST, SUCCESS, FAILURE } from 'middlewares/requests-middleware';

const NULL_FETCH_REQUEST: FetchRequest = {
  isPending: false,
  isSuccessful: false,
  isFailure: false,
};

const requestState = 'Request';

export function getRequestName(type: string): RequestType {
  const normalizedType = type
    .replace(REQUEST, '')
    .replace(SUCCESS, '')
    .replace(FAILURE, '');

  return camelCase(normalizedType) + requestState;
}

export default function requests(state: Requests = {}, action: FSA): Requests {
  const { type, meta } = action;

  switch (type) {
    case RESET_ALL_STATE:
      return {};

    case RESET_REQUEST_STATE: {
      const newState: Requests = {};
      action.payload.forEach((domain) => {
        newState[getRequestName(domain)] = NULL_FETCH_REQUEST;
      });

      return { ...state, ...newState };
    }

    default: // Fallthrough
  }

  // requestStatus is a field specially designed and owned by api request actions
  if (!meta || !meta.requestStatus) return state;
  switch (meta.requestStatus) {
    case REQUEST:
      return {
        ...state,
        [getRequestName(type)]: {
          ...NULL_FETCH_REQUEST,
          isPending: true,
        },
      };

    case SUCCESS:
      return {
        ...state,
        [getRequestName(type)]: {
          ...NULL_FETCH_REQUEST,
          isSuccessful: true,
        },
      };

    case FAILURE:
      return {
        ...state,
        [getRequestName(type)]: {
          ...NULL_FETCH_REQUEST,
          isFailure: true,
          error: action.payload,
        },
      };

    default:
      return state;
  }
}
