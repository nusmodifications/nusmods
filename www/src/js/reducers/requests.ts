// @flow
import type { FSA } from 'types/redux';
import type { Requests } from 'types/reducers';

import { API_REQUEST } from 'actions/requests';
import { FAILURE, REQUEST, SUCCESS } from 'types/reducers';

export default function requests(state: Requests = {}, action: FSA): Requests {
  const { meta } = action;

  // requestStatus is a field specially designed and owned by api request actions
  if (!meta || !meta.requestStatus || !meta[API_REQUEST]) {
    return state;
  }

  const key = meta[API_REQUEST];

  switch (meta.requestStatus) {
    case REQUEST:
      return {
        ...state,
        [key]: {
          status: REQUEST,
        },
      };

    case SUCCESS:
      return {
        ...state,
        [key]: {
          status: SUCCESS,
        },
      };

    case FAILURE:
      return {
        ...state,
        [key]: {
          status: FAILURE,
          error: action.payload,
        },
      };

    default:
      return state;
  }
}
