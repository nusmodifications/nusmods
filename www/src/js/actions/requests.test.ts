import { requestAction } from './requests';

describe(requestAction, () => {
  it('should use type as key if two params are provided', () => {
    const action = requestAction('TEST_ACTION', {
      url: 'http://api.example.com',
    });

    expect(action.type).toEqual('TEST_ACTION');
    expect(action.meta.API_REQUEST).toEqual('TEST_ACTION');

    expect(action).toMatchSnapshot();
  });

  it('should use provided key if three params are provided', () => {
    const action = requestAction('TEST_KEY', 'TEST_ACTION', {
      url: 'http://api.example.com',
    });

    expect(action.type).toEqual('TEST_ACTION');
    expect(action.meta.API_REQUEST).toEqual('TEST_KEY');

    expect(action).toMatchSnapshot();
  });
});
