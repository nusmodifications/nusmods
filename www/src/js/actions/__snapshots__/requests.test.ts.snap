// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requestAction should use provided key if three params are provided 1`] = `
Object {
  "meta": Object {
    "API_REQUEST": "TEST_KEY",
  },
  "payload": Object {
    "url": "http://api.example.com",
  },
  "type": "TEST_ACTION",
}
`;

exports[`requestAction should use type as key if two params are provided 1`] = `
Object {
  "meta": Object {
    "API_REQUEST": "TEST_ACTION",
  },
  "payload": Object {
    "url": "http://api.example.com",
  },
  "type": "TEST_ACTION",
}
`;
